x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtSignatureValidatorTests.java

```diff
@@ -0,0 +1,267 @@
+        });
+
+        Thread t2 = new Thread(() -> {
+            safeAwait(barrier);
+            signatureValidator.validate(randomIdentifier(), signedJWT, primaryListener);
+        });
+
+        t1.start();
+        t2.start();
+        safeAwait(barrier); // kick off the work
+        t1.join();
+        t2.join();
+
+        /**
+         * THESE VALUES REPRESENT HOW IT ACTUALLY WORKS, NOT HOW IT SHOULD WORK
+         * DUE THE BUG THE SECOND VALIDATION INCORRECTLY FAILS
+         */
+        assertThat(validateSignatureAttemptCounter.get(), is(3)); // should be 4, but the bug results in 1
+        assertThat(finalSuccessCounter.get(), is(1)); // should be 2, but the bug results in 1
+        assertThat(finalFailureCounter.get(), is(1)); // should be 0, but the bug results in 1
+        assertThat(reloadAttemptCounter.get(), is(2)); // 2 is correct with or without the bug
+    }
+
+    public void testJwtSignVerifyPassedForAllSupportedAlgorithms() {
+        // Pass: "ES256", "ES384", "ES512", RS256", "RS384", "RS512", "PS256", "PS384", "PS512, "HS256", "HS384", "HS512"
+        for (final String signatureAlgorithm : JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS) {
+            try {
+                helpTestSignatureAlgorithm(signatureAlgorithm, false);
+            } catch (Exception e) {
+                fail("signature validation with algorithm [" + signatureAlgorithm + "] should have succeeded");
+            }
+        }
+        // Fail: "ES256K"
+        final Exception exp1 = expectThrows(JOSEException.class, () -> helpTestSignatureAlgorithm(JWSAlgorithm.ES256K.getName(), false));
+        final String msg1 = "Unsupported signature algorithm ["
+            + JWSAlgorithm.ES256K
+            + "]. Supported signature algorithms are "
+            + JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS
+            + ".";
+        assertThat(exp1.getMessage(), equalTo(msg1));
+    }
+
+    private void helpTestSignatureAlgorithm(final String signatureAlgorithm, final boolean requireOidcSafe) throws Exception {
+        logger.trace("Testing signature algorithm " + signatureAlgorithm);
+        final JWK jwk = JwtTestCase.randomJwk(signatureAlgorithm, requireOidcSafe);
+        final SecureString serializedJWTOriginal = JwtTestCase.randomBespokeJwt(jwk, signatureAlgorithm);
+        final SignedJWT parsedSignedJWT = SignedJWT.parse(serializedJWTOriginal.toString());
+        // OK ... so this empty jwtSignatureValidator is weird.
+        // It's because there once was a static method that held the validateSignature method. It could be invoked independent of which
+        // subclass was actually invoked. To help with test a bug with the signature validation the static validateSignature was moved
+        // from a static method to default method in the interface. This is not ideal, but neither was un-testable code.
+        JwtSignatureValidator jwtSignatureValidator = (tokenPrincipal, jwt, listener) -> {};
```

Comment:
creating a dummy implementation is the only change here and is explained in the comment.